Adjust max. brightness and saturation for Hue API

dominicdesu opened the pull request:
Set maximum values for brightness and saturation to comply with the
official Hue API (254 instead of 255).
Extend HueLightHandlerOSGiTest test to include Increase/Decrease type.
Fix two tests of HueLightHandlerOSGiTest where incorrect channel was used.

Fixes https://bugs.eclipse.org/bugs/show_bug.cgi?id=470117

Signed-off-by: Dominic Lerbs <dominicdesu@hallojapan.de>

dnobel commented:
@tomhoefer Can you please have a look at this PR? Isn´t this something the Jue library should take care of?

tomhoefer commented:
@de I have fixed this in https://github.com/qivicon/Jue/commit/a9549a707ec59fdb947d7f45def90367db38bf8e which corresponds to the current Jue Library in ESH

tomhoefer commented:
But the PR is also fine for me ... So no need to handle IllegalArgumentException

dominicdesu commented:
I agree, the Jue library used by ESH was already adjusted to comply with the Hue API. However it was still possible in the Hue binding to send an invalid value (255) to the Jue library which would "answer" with an IllegalArgumentException. This has been fixed with the pull request.

kaikreuzer commented:
Thanks @dominicdesu, the code looks good and is much better readable than before!

Just one question about the color temperature. According to the Hue documentation, the "ct" value is defined in https://en.wikipedia.org/wiki/Mired and thus does not have a range of 0-254. The code to handle INCREASE/DECREASE seems to be completely wrong (no complaint about you since this was like this before this way already). When sending DECREASE commands to the color temp channel, I also get
```
2015-06-17 10:50:04 [INFO ] [runtime.busevents             :22   ] - hue_LCT001_001788142732_3_color_temperature received command DECREASE
2015-06-17 10:50:04 [DEBUG] [.c.thing.internal.ThingManager:222  ] - Delegating command 'DECREASE' for item 'hue_LCT001_001788142732_3_color_temperature' to handler for channel 'hue:LCT001:001788142732:3:color_temperature'
2015-06-17 10:50:04 [ERROR] [.c.thing.internal.ThingManager:227  ] - Exception occured while calling handler: Color temperature out of rangejava.lang.IllegalArgumentException: Color temperature out of range
	at nl.q42.jue.StateUpdate.setColorTemperature(StateUpdate.java:129)
	at org.eclipse.smarthome.binding.hue.handler.HueLightHandler.handleCommand(HueLightHandler.java:168)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.receiveCommand(ThingManager.java:225)
```
which confirms my assumption.

Would you be willing to address this as well?
Another observation is that the lastSentColorTemp and lastSentBrightness fields are not set when handling OnOffType and PercentType commands. This is also a bug that should be addressed.

So you see, enough to do :-) Would be great if you could have a look at these. I will also soon answer your question in the forum about the additional functionalities.

dominicdesu commented:
Thanks for the thorough review, @kaikreuzer. This looks indeed quite incorrect. I'll have a look at your remarks and push another commit as soon as fixed!

dominicdesu commented:
Hi,
I tried to fetch and merge the latest changes from esh/master. Then I pushed my changes to github and now all the merged commits appear in this pull request. Could somebody advice me what I did wrong and how I can fix it?

dnobel commented:
Looks like you rebased the commits. I would suggest to create a new branch based on the master, cherry-pick your commit and then do a force push to the remote branch.

dominicdesu commented:
@dnobel yes I was using the rebase command, didn't know it would create such a mess. I followed your advise and now it looks much better. Thanks!

Commit 356da2b fixes the Increase/Decrease command for the color channel.
Commit 83dfd00 adds more tests and refactors the test class so that various constellations can be tested easier.

I have not changed the caching (lastSent...) for the OnOffType and PercentType for two reasons:

1. The last sent value is not directly relevant for the OnOffType and PercentType because they send a fixed value. Only the IncreaseDecreaseType needs the last sent value to determine the next value it has to sent. So the only situation when the last sent value is necessary for OnOffType and PercentType is when immediately afterwards an IncreaseDecreaseType command is performed.

2. A clean implementation would require quite some refactoring of the hue binding in my opinion. Let's take for example a change of the brightness: The `StateUpdate` containing the new brightness is created within the static methods of the `LightStateConverter` and then returned to the `HueLightHandler`. The `StateUpdate` will be sent to the bridge and the lastSentBrightness has to be set. However, `StateUpdate` has no getter methods. There is no possibility in `HueLightHandler` to get the new brightness value which has been sent. I could think of some *quick hacks* to circumvent this problem but they would only make the code more complex. Instead of hacking a solution for a very minor problem (see point 1) it would be nicer to try to improve the OO-design of the Hue binding. Or maybe you have some other idea?

If you agree, I would prefer to create a seperate ticket for the lastSent issue, as a) the refactoring would change quite some code, b) it might need more discussion and c) I would like to work on something else first :-)

dominicdesu commented:
Hi,
is here anything open from my side? Just let me know if this should be implemented differently or have any concerns and I'll look into it.

dnobel commented:
Sorry, I didn´t realized that it is assigned to me. Can you please explain me again, what has been fixed? Is there a bugzilla issue?

> Commit 356da2b fixes the Increase/Decrease command for the color channel.

What was wrong with increase/decrease and how does it work now?

I also see that you refactored the whole conversation from the smarthome types into the jue library types. Did you change anything else? In the original description you wrote something about the max value for hue (254 vs 255) Is also fixed?

Thanks in advance!

dominicdesu commented:
Hi,

let me first explain what was fixed/has changed and then answer your questions.

This pull request fixed two issues:
### Incorrect max brightness
ESH used incorrect max brightness value of 255 (official Hue maximum is 254) in case a IncreaseDecrease command was used on the color channel. This, including some refactoring as well as additions and fixes for the test classes, was the initial pull request in commit b405206 (main) and 2267947 (test).

### Incorrect color temperature range
ESH used incorrect color temperature range of 0-254 (official Hue is 153-500) in case a IncreaseDecrease command was used on the Mired color temperature channel as mentioned by @kaikreuzer in the code review. This was commit 356da2b with the following changes:
* The main change for the color temperature is in [line 131 of HueLightHandler](https://github.com/dominicdesu/smarthome/commit/356da2b94e50094e3091020363f8b6fc31183d01#diff-2ba4d1bdb2c89e2c8b66b342a88ed23eR131).
* Moved the calculateNewBrightness() method from HueLightHandler to LightStateConverter 
* Some smaller refactoring to make the code more concise / easier understandable

Last, I have refactored the test (before it was not possible to modify the expected reply from the mocked HueBridge) and added more test cases, especially for the IncreaseDecrease command in commit 83dfd00. Unfortunately the formatting was automatically adjusted by Eclipse and therefore git compare is difficult.

If you want, I can revert some of the refactorings from commit 356da2b to reduce the number of changed lines. However I believe that these refactorings improve some misleading naming and also help to easier understand the code.


Now about your remaining questions: 
> Is there a bugzilla issue?

The bugzilla issue is https://bugs.eclipse.org/bugs/show_bug.cgi?id=470117. This only includes the first fixed issue as the second one came up during the code review of this pull request.

> I also see that you refactored the whole conversation from the smarthome types into the jue library types. 

This might be a controversial change. The conversions seemed to be more complex as they need to be. I tried to simplify them. Looking at it now, I would have probably left them unchanged. Though I still don't see where they could be wrong, I can revert if you prefer.

> Did you change anything else?

* I replaced `COLOR_TEMPERATURE_FACTOR` (3.46) and `COLOR_TEMPERATURE_OFFSET` (154) with `MIN_COLOR_TEMPERATURE` (153) and `MAX_COLOR_TEMPERATURE` (500) because factor and offset made it a little bit hard to understand (at least in my opinion) it is actually a range.
* Also renamed some methods to better reflect what they are doing

I did *not* change anything about the caching of the last sent values as mentioned above.

I hope all other questions were answered with the explanation of the changes. Let me know if anything is unclear or anything should be changed!

dnobel commented:
OK, thanks for the explanation, think I understand it now :).

Before I can merge it, can you please squash your 4 commits into 1 with a Signed-off-by line at the end. Because your third and fourth commit don´t have a Signed-off-by.

Thanks.

dominicdesu commented:
done!

dnobel commented:
Thanks.

dominicdesu commented:
Hi, as it seems [some tests failed](https://hudson.eclipse.org/smarthome/job/SmartHomeDistribution/538/testReport/org.eclipse.smarthome.binding.hue.test/)... I will have a look at it today evening...

dnobel commented:
Yes, it seems to be related with your PR. 19 hue tests fail. Did they run on your PC?

dominicdesu commented:
@dnobel They did run on my PC. At least until, just before making the final commit, I thought I could simplify a variable return in HueLightState.groovy (see #432 for details).

I have created an additional pull request #432 which fixes this and another issue. Let me know if this should be handled differently.

In the future, I'll make sure to execute the tests once again after the final changes (even if it's only comments).
